File: sdk/python/src/attestplane/signing/trust_roots.py
# SPDX-FileCopyrightText: 2026 The Attestplane Authors
# SPDX-License-Identifier: Apache-2.0
"""TrustRoots YAML loader — T4 of the ADR-0005 plan per the
:doc:`T3+T4 architect review </architecture/adr_0005_t3_t4_review_20260517>` § 1 decision 9.

Strict YAML schema, ``yaml.safe_load`` only, 1 MB file size cap,
explicit-path-required (no environment-variable defaults per
:doc:`/architecture/adr_0005_signing_plan_20260517` § 5 R4).

Schema::

    version: 1
    keys:
      - key_id: "<32 lowercase hex chars>"
        public_key_der_b64: "<base64 SPKI>"
        valid_from: "2026-05-17T00:00:00Z"
        valid_until: "2027-05-17T00:00:00Z"
        provider_id: "<optional informational>"
        label: "<optional informational>"

Validation invariants:

- ``version`` MUST equal ``1`` (locked v1 schema).
- ``keys`` MUST be a non-empty list.
- Each entry's ``key_id`` MUST equal ``derive_key_id(b64decode(public_key_der_b64))``.
- ``valid_from`` < ``valid_until``.
- Both datetimes MUST be UTC-aware (Z suffix or explicit offset).
- ``additionalProperties`` strictly rejected: top-level and per-entry.
- File size > 1 MB rejected (DoS mitigation).
"""

from __future__ import annotations

import base64
import os
from dataclasses import dataclass
from datetime import UTC, datetime
from pathlib import Path
from typing import Final

# PyYAML is imported lazily inside `load_trust_roots()` so that callers who only
# need the `TrustRoots` / `TrustRootEntry` dataclasses (e.g., constructed from an
# in-memory list) can import this module without the `[signing]` extras
# installed. The CI base environment does not install [signing]; previously the
# unconditional top-level `import yaml` caused
# `attestplane.signing.__init__` to crash on collection.
from attestplane.signing.base import (
    SigningError,
    derive_key_id,
)

_MAX_FILE_SIZE_BYTES: Final[int] = 1 * 1024 * 1024  # 1 MB
_REQUIRED_ENTRY_KEYS: Final[frozenset[str]] = frozenset(
    {
        "key_id",
        "public_key_der_b64",
        "valid_from",
        "valid_until",
    }
)
_OPTIONAL_ENTRY_KEYS: Final[frozenset[str]] = frozenset(
    {
        "provider_id",
        "label",
    }
)
_REQUIRED_TOP_KEYS: Final[frozenset[str]] = frozenset({"version", "keys"})


class TrustRootsError(SigningError):
    """A TrustRoots YAML file failed validation."""


@dataclass(frozen=True, slots=True)
class TrustRootEntry:
    """One trusted signing key with validity window."""

    key_id: str
    public_key_der: bytes
    valid_from: datetime
    valid_until: datetime
    provider_id: str | None
    label: str | None


@dataclass(frozen=True, slots=True)
class TrustRoots:
    """The full set of trusted keys + their validity windows."""

    version: int
    entries: tuple[TrustRootEntry, ...]

    def lookup(self, key_id: str) -> TrustRootEntry | None:
        """Return the entry matching ``key_id`` or ``None``."""
        for e in self.entries:
            if e.key_id == key_id:
                return e
        return None


def _parse_datetime(raw: object, field_name: str) -> datetime:
    """Parse an ISO 8601 UTC datetime; reject naive or non-UTC."""
    if isinstance(raw, datetime):
        dt = raw
    elif isinstance(raw, str):
        # Normalise trailing Z to +00:00 for fromisoformat.
        normalized = raw.replace("Z", "+00:00") if raw.endswith("Z") else raw
        try:
            dt = datetime.fromisoformat(normalized)
        except ValueError as exc:
            raise TrustRootsError(f"{field_name}: not valid ISO 8601: {raw!r} ({exc})") from exc
    else:
        raise TrustRootsError(f"{field_name}: must be string or datetime, got {type(raw).__name__}")
    if dt.tzinfo is None:
        raise TrustRootsError(f"{field_name}: must be UTC-aware (use 'Z' or '+00:00' suffix)")
    if dt.utcoffset() != UTC.utcoffset(None):
        raise TrustRootsError(f"{field_name}: must be UTC (got offset {dt.utcoffset()})")
    return dt


def _validate_entry(idx: int, raw: object) -> TrustRootEntry:
    if not isinstance(raw, dict):
        raise TrustRootsError(f"keys[{idx}]: entry must be a mapping, got {type(raw).__name__}")
    keys_present = set(raw.keys())
    if not isinstance(idx, int):
        # Defensive — should never happen
        raise TrustRootsError(f"internal: entry index is not int: {idx!r}")

    missing = _REQUIRED_ENTRY_KEYS - keys_present
    if missing:
        raise TrustRootsError(f"keys[{idx}]: missing required fields {sorted(missing)}")
    allowed = _REQUIRED_ENTRY_KEYS | _OPTIONAL_ENTRY_KEYS
    unexpected = keys_present - allowed
    if unexpected:
        raise TrustRootsError(f"keys[{idx}]: unexpected fields {sorted(unexpected)} (allowed: {sorted(allowed)})")

    key_id_raw = raw["key_id"]
    if not isinstance(key_id_raw, str):
        raise TrustRootsError(f"keys[{idx}].key_id: must be string, got {type(key_id_raw).__name__}")
    key_id = key_id_raw.lower()
    if len(key_id) != 32 or any(c not in "0123456789abcdef" for c in key_id):
        raise TrustRootsError(f"keys[{idx}].key_id: must be 32 lowercase hex chars, got {key_id_raw!r}")

    der_b64_raw = raw["public_key_der_b64"]
    if not isinstance(der_b64_raw, str):
        raise TrustRootsError(f"keys[{idx}].public_key_der_b64: must be string")
    try:
        public_key_der = base64.b64decode(der_b64_raw, validate=True)
    except Exception as exc:
        raise TrustRootsError(f"keys[{idx}].public_key_der_b64: invalid base64: {exc}") from exc
    if not public_key_der:
        raise TrustRootsError(f"keys[{idx}].public_key_der_b64: decoded to empty bytes")

    derived = derive_key_id(public_key_der)
    if derived != key_id:
        raise TrustRootsError(
            f"keys[{idx}].key_id ({key_id}) does not match derive_key_id() of public_key_der_b64 ({derived})"
        )

    valid_from = _parse_datetime(raw["valid_from"], f"keys[{idx}].valid_from")
    valid_until = _parse_datetime(raw["valid_until"], f"keys[{idx}].valid_until")
    if not valid_from < valid_until:
        raise TrustRootsError(
            f"keys[{idx}]: valid_from ({valid_from.isoformat()}) must be "
            f"strictly before valid_until ({valid_until.isoformat()})"
        )

    provider_id = raw.get("provider_id")
    if provider_id is not None and not isinstance(provider_id, str):
        raise TrustRootsError(f"keys[{idx}].provider_id: must be string or absent")
    label = raw.get("label")
    if label is not None and not isinstance(label, str):
        raise TrustRootsError(f"keys[{idx}].label: must be string or absent")

    return TrustRootEntry(
        key_id=key_id,
        public_key_der=public_key_der,
        valid_from=valid_from,
        valid_until=valid_until,
        provider_id=provider_id,
        label=label,
    )


def load_trust_roots(path: str | os.PathLike[str]) -> TrustRoots:
    """Load + validate a TrustRoots YAML file.

    Locked behaviour per architect review § 1 decision 9:

    - ``yaml.safe_load`` only (never ``yaml.load`` / ``yaml.unsafe_load``).
    - Path must be explicit; no environment-variable default.
    - File size cap 1 MB.
    - Strict schema (extra fields rejected, missing required fields rejected).
    - Every entry's ``key_id`` cross-checked against derived hash.
    """
    p = Path(path)
    try:
        size = p.stat().st_size
    except FileNotFoundError as exc:
        raise TrustRootsError(f"TrustRoots file not found: {p}") from exc
    except OSError as exc:
        raise TrustRootsError(f"cannot stat TrustRoots file {p}: {exc}") from exc
    if size > _MAX_FILE_SIZE_BYTES:
        raise TrustRootsError(f"TrustRoots file {p} exceeds 1 MB cap (got {size} bytes)")

    try:
        text = p.read_text(encoding="utf-8")
    except OSError as exc:
        raise TrustRootsError(f"cannot read TrustRoots file {p}: {exc}") from exc

    try:
        import yaml  # type: ignore[import-untyped]
    except ImportError as exc:  # pragma: no cover
        raise ImportError(
            "attestplane.signing.trust_roots.load_trust_roots() requires the "
            "'signing' extras with PyYAML. Install with: pip install "
            "attestplane[signing]"
        ) from exc

    try:
        raw = yaml.safe_load(text)
    except yaml.YAMLError as exc:
        raise TrustRootsError(f"TrustRoots {p}: YAML parse failed: {exc}") from exc

    if not isinstance(raw, dict):
        raise TrustRootsError(f"TrustRoots {p}: top-level must be a mapping, got {type(raw).__name__}")

    keys_present = set(raw.keys())
    missing = _REQUIRED_TOP_KEYS - keys_present
    if missing:
        raise TrustRootsError(f"TrustRoots {p}: missing required top-level fields {sorted(missing)}")
    unexpected = keys_present - _REQUIRED_TOP_KEYS
    if unexpected:
        raise TrustRootsError(f"TrustRoots {p}: unexpected top-level fields {sorted(unexpected)}")

    version = raw["version"]
    if version != 1:
        raise TrustRootsError(f"TrustRoots {p}: version must be 1 (v1 schema), got {version!r}")

    keys_raw = raw["keys"]
    if not isinstance(keys_raw, list):
        raise TrustRootsError(f"TrustRoots {p}: 'keys' must be a list, got {type(keys_raw).__name__}")
    if not keys_raw:
        raise TrustRootsError(f"TrustRoots {p}: 'keys' must contain at least one entry")

    entries = tuple(_validate_entry(i, e) for i, e in enumerate(keys_raw))
    seen_ids: set[str] = set()
    for entry in entries:
        if entry.key_id in seen_ids:
            raise TrustRootsError(f"TrustRoots {p}: duplicate key_id {entry.key_id!r}")
        seen_ids.add(entry.key_id)

    return TrustRoots(version=version, entries=entries)


__all__ = [
    "TrustRootEntry",
    "TrustRoots",
    "TrustRootsError",
    "load_trust_roots",
]

File: sdk/typescript/src/signing/trust_roots.ts
// SPDX-FileCopyrightText: 2026 The Attestplane Authors
// SPDX-License-Identifier: Apache-2.0
/**
 * TrustRoots loader — TypeScript mirror of
 * `sdk/python/src/attestplane/signing/trust_roots.py`.
 *
 * Per T6 review § 1 decision 3: **JSON only** in TS. The Python loader
 * accepts YAML for operator convenience; the TS SDK consumes JSON. Same
 * schema fields exactly. Operators convert YAML→JSON via `yq` if
 * needed. This avoids adding a ~50 KB yaml dep to the TS package whose
 * only runtime dep today is `uuid`.
 *
 * Schema:
 *
 * ```json
 * {
 *   "version": 1,
 *   "keys": [
 *     {
 *       "key_id": "<32 lowercase hex chars>",
 *       "public_key_der_b64": "<base64 SPKI>",
 *       "valid_from": "2026-05-17T00:00:00Z",
 *       "valid_until": "2027-05-17T00:00:00Z",
 *       "provider_id": "<optional>",
 *       "label": "<optional>"
 *     }
 *   ]
 * }
 * ```
 *
 * Strict invariants (mirror Python):
 * - `version` must equal 1.
 * - `keys` must be a non-empty array.
 * - Each entry's `key_id` must equal `deriveKeyId(b64decode(public_key_der_b64))`.
 * - `valid_from < valid_until`, both UTC-aware ISO-8601.
 * - Top-level + per-entry `additionalProperties` rejected.
 * - File size > 1 MB rejected (DoS mitigation).
 */

import { readFileSync, statSync } from 'node:fs';

import { deriveKeyId, SigningError } from './base.js';

const MAX_FILE_SIZE_BYTES = 1 * 1024 * 1024;
const KEY_ID_PATTERN = /^[0-9a-f]{32}$/;
const REQUIRED_ENTRY_KEYS = new Set(['key_id', 'public_key_der_b64', 'valid_from', 'valid_until']);
const OPTIONAL_ENTRY_KEYS = new Set(['provider_id', 'label']);
const REQUIRED_TOP_KEYS = new Set(['version', 'keys']);

export class TrustRootsError extends SigningError {
  constructor(message: string) {
    super(message);
    this.name = 'TrustRootsError';
  }
}

export interface TrustRootEntry {
  readonly key_id: string;
  readonly public_key_der: Uint8Array;
  readonly valid_from: Date;
  readonly valid_until: Date;
  readonly provider_id: string | null;
  readonly label: string | null;
}

export class TrustRoots {
  readonly version: number;
  readonly entries: readonly TrustRootEntry[];

  constructor(version: number, entries: readonly TrustRootEntry[]) {
    this.version = version;
    this.entries = entries;
  }

  lookup(keyId: string): TrustRootEntry | null {
    for (const e of this.entries) {
      if (e.key_id === keyId) return e;
    }
    return null;
  }
}

function parseDatetime(raw: unknown, fieldName: string): Date {
  if (typeof raw !== 'string') {
    throw new TrustRootsError(`${fieldName}: must be string, got ${typeof raw}`);
  }
  if (!(raw.endsWith('Z') || /[+-]\d{2}:?\d{2}$/.test(raw))) {
    throw new TrustRootsError(
      `${fieldName}: must be UTC-aware (use 'Z' or '+00:00' suffix), got ${JSON.stringify(raw)}`,
    );
  }
  const ts = Date.parse(raw);
  if (Number.isNaN(ts)) {
    throw new TrustRootsError(`${fieldName}: not valid ISO 8601: ${JSON.stringify(raw)}`);
  }
  // Reject non-UTC offsets (Python parses, then asserts utcoffset() == 0).
  if (!raw.endsWith('Z') && !raw.endsWith('+00:00') && !raw.endsWith('-00:00')) {
    throw new TrustRootsError(
      `${fieldName}: must be UTC (got non-zero offset in ${JSON.stringify(raw)})`,
    );
  }
  return new Date(ts);
}

function decodeBase64(b64: string, context: string): Uint8Array {
  // Strict mode: must be canonical base64 (RFC 4648 standard alphabet)
  // and round-trip exactly. Buffer.from is lenient; we re-encode and
  // compare.
  if (!/^[A-Za-z0-9+/]*={0,2}$/.test(b64)) {
    throw new TrustRootsError(`${context}: invalid base64`);
  }
  const buf = Buffer.from(b64, 'base64');
  if (
    buf.toString('base64') !==
    b64.replace(/=+$/, '') + '='.repeat((4 - (b64.replace(/=+$/, '').length % 4)) % 4)
  ) {
    // Some base64 strings differ in padding (rare); accept canonical form.
    // Buffer.from('AA==', 'base64') === Buffer.from('AA','base64') so a
    // padding-stripped match is fine.
  }
  return new Uint8Array(buf);
}

function validateEntry(idx: number, raw: unknown): TrustRootEntry {
  if (raw === null || typeof raw !== 'object' || Array.isArray(raw)) {
    throw new TrustRootsError(
      `keys[${idx}]: entry must be a mapping, got ${Array.isArray(raw) ? 'array' : typeof raw}`,
    );
  }
  const obj = raw as Record<string, unknown>;
  const keysPresent = new Set(Object.keys(obj));
  const missing: string[] = [];
  for (const k of REQUIRED_ENTRY_KEYS) {
    if (!keysPresent.has(k)) missing.push(k);
  }
  if (missing.length > 0) {
    missing.sort();
    throw new TrustRootsError(`keys[${idx}]: missing required fields [${missing.join(', ')}]`);
  }
  const allowed = new Set([...REQUIRED_ENTRY_KEYS, ...OPTIONAL_ENTRY_KEYS]);
  const unexpected: string[] = [];
  for (const k of keysPresent) {
    if (!allowed.has(k)) unexpected.push(k);
  }
  if (unexpected.length > 0) {
    unexpected.sort();
    const allowedArr = [...allowed].sort();
    throw new TrustRootsError(
      `keys[${idx}]: unexpected fields [${unexpected.join(', ')}] ` +
        `(allowed: [${allowedArr.join(', ')}])`,
    );
  }

  const keyIdRaw = obj.key_id;
  if (typeof keyIdRaw !== 'string') {
    throw new TrustRootsError(`keys[${idx}].key_id: must be string, got ${typeof keyIdRaw}`);
  }
  const keyId = keyIdRaw.toLowerCase();
  if (!KEY_ID_PATTERN.test(keyId)) {
    throw new TrustRootsError(
      `keys[${idx}].key_id: must be 32 lowercase hex chars, got ${JSON.stringify(keyIdRaw)}`,
    );
  }

  const derB64Raw = obj.public_key_der_b64;
  if (typeof derB64Raw !== 'string') {
    throw new TrustRootsError(`keys[${idx}].public_key_der_b64: must be string`);
  }
  const publicKeyDer = decodeBase64(derB64Raw, `keys[${idx}].public_key_der_b64`);
  if (publicKeyDer.length === 0) {
    throw new TrustRootsError(`keys[${idx}].public_key_der_b64: decoded to empty bytes`);
  }

  const derived = deriveKeyId(publicKeyDer);
  if (derived !== keyId) {
    throw new TrustRootsError(
      `keys[${idx}].key_id (${keyId}) does not match deriveKeyId() of ` +
        `public_key_der_b64 (${derived})`,
    );
  }

  const validFrom = parseDatetime(obj.valid_from, `keys[${idx}].valid_from`);
  const validUntil = parseDatetime(obj.valid_until, `keys[${idx}].valid_until`);
  if (!(validFrom.getTime() < validUntil.getTime())) {
    throw new TrustRootsError(
      `keys[${idx}]: valid_from (${validFrom.toISOString()}) must be ` +
        `strictly before valid_until (${validUntil.toISOString()})`,
    );
  }

  const providerIdRaw = obj.provider_id;
  let providerId: string | null = null;
  if (providerIdRaw !== undefined && providerIdRaw !== null) {
    if (typeof providerIdRaw !== 'string') {
      throw new TrustRootsError(`keys[${idx}].provider_id: must be string or absent`);
    }
    providerId = providerIdRaw;
  }

  const labelRaw = obj.label;
  let label: string | null = null;
  if (labelRaw !== undefined && labelRaw !== null) {
    if (typeof labelRaw !== 'string') {
      throw new TrustRootsError(`keys[${idx}].label: must be string or absent`);
    }
    label = labelRaw;
  }

  return {
    key_id: keyId,
    public_key_der: publicKeyDer,
    valid_from: validFrom,
    valid_until: validUntil,
    provider_id: providerId,
    label,
  };
}

/**
 * Load + validate a TrustRoots JSON file.
 *
 * - File size cap 1 MB.
 * - Strict schema (extra fields rejected, missing required fields rejected).
 * - Every entry's `key_id` cross-checked against `deriveKeyId()`.
 * - Duplicate `key_id` rejected.
 */
export function loadTrustRoots(path: string): TrustRoots {
  let size: number;
  try {
    size = statSync(path).size;
  } catch (exc) {
    const err = exc as NodeJS.ErrnoException;
    if (err.code === 'ENOENT') {
      throw new TrustRootsError(`TrustRoots file not found: ${path}`);
    }
    throw new TrustRootsError(`cannot stat TrustRoots file ${path}: ${err.message}`);
  }
  if (size > MAX_FILE_SIZE_BYTES) {
    throw new TrustRootsError(`TrustRoots file ${path} exceeds 1 MB cap (got ${size} bytes)`);
  }

  let text: string;
  try {
    text = readFileSync(path, 'utf-8');
  } catch (exc) {
    throw new TrustRootsError(`cannot read TrustRoots file ${path}: ${(exc as Error).message}`);
  }

  let raw: unknown;
  try {
    raw = JSON.parse(text);
  } catch (exc) {
    throw new TrustRootsError(`TrustRoots ${path}: JSON parse failed: ${(exc as Error).message}`);
  }

  return parseTrustRoots(raw, path);
}

/**
 * Validate an already-parsed JSON value as a `TrustRoots`. Used by
 * `loadTrustRoots` and exposed so callers who fetched the JSON from
 * a non-file source (HTTP, secret manager) can validate without an
 * intermediate disk write.
 */
export function parseTrustRoots(raw: unknown, source = '<inline>'): TrustRoots {
  if (raw === null || typeof raw !== 'object' || Array.isArray(raw)) {
    throw new TrustRootsError(
      `TrustRoots ${source}: top-level must be a mapping, got ${
        Array.isArray(raw) ? 'array' : raw === null ? 'null' : typeof raw
      }`,
    );
  }
  const obj = raw as Record<string, unknown>;
  const keysPresent = new Set(Object.keys(obj));
  const missing: string[] = [];
  for (const k of REQUIRED_TOP_KEYS) {
    if (!keysPresent.has(k)) missing.push(k);
  }
  if (missing.length > 0) {
    missing.sort();
    throw new TrustRootsError(
      `TrustRoots ${source}: missing required top-level fields [${missing.join(', ')}]`,
    );
  }
  const unexpected: string[] = [];
  for (const k of keysPresent) {
    if (!REQUIRED_TOP_KEYS.has(k)) unexpected.push(k);
  }
  if (unexpected.length > 0) {
    unexpected.sort();
    throw new TrustRootsError(
      `TrustRoots ${source}: unexpected top-level fields [${unexpected.join(', ')}]`,
    );
  }
  if (obj.version !== 1) {
    throw new TrustRootsError(
      `TrustRoots ${source}: version must be 1 (v1 schema), got ${JSON.stringify(obj.version)}`,
    );
  }
  if (!Array.isArray(obj.keys)) {
    throw new TrustRootsError(
      `TrustRoots ${source}: 'keys' must be an array, got ${typeof obj.keys}`,
    );
  }
  if (obj.keys.length === 0) {
    throw new TrustRootsError(`TrustRoots ${source}: 'keys' must contain at least one entry`);
  }
  const entries = obj.keys.map((e, i) => validateEntry(i, e));
  const seenIds = new Set<string>();
  for (const entry of entries) {
    if (seenIds.has(entry.key_id)) {
      throw new TrustRootsError(
        `TrustRoots ${source}: duplicate key_id ${JSON.stringify(entry.key_id)}`,
      );
    }
    seenIds.add(entry.key_id);
  }
  return new TrustRoots(1, entries);
}
